### PR TITLE
Optimize `Log10()`

### PR DESCRIPTION
**Pull request by fyfyrchik**
Hello, thanks for the awesome library!

I was looking for something to optimize, so here it is:
Bit twiddling hacks page has simple algorithm for calculating log10 of an integer:
https://graphics.stanford.edu/~seander/bithacks.html#IntegerLog10

Implement it here.

 ```
    goos: linux
    goarch: amd64
    pkg: github.com/holiman/uint256
    cpu: 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
                          │     old      │                 new                 │
                          │    sec/op    │   sec/op     vs base                │
    Log10/Log10/uint256-8   1016.5n ± 6%   762.9n ± 1%  -24.95% (p=0.000 n=10)
    Log10/Log10/big-8        27.66µ ± 5%   28.56µ ± 5%        ~ (p=0.393 n=10)
    geomean                  5.303µ        4.668µ       -11.97%

                          │      old       │                  new                  │
                          │      B/op      │     B/op      vs base                 │
    Log10/Log10/uint256-8     0.000 ± 0%       0.000 ± 0%       ~ (p=1.000 n=10) ¹
    Log10/Log10/big-8       22.88Ki ± 0%     22.88Ki ± 0%       ~ (p=1.000 n=10) ¹
    geomean                              ²                 +0.00%                ²
    ¹ all samples are equal
    ² summaries must be >0 to compute geomean

                          │     old      │                 new                 │
                          │  allocs/op   │ allocs/op   vs base                 │
    Log10/Log10/uint256-8   0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
    Log10/Log10/big-8       510.0 ± 0%     510.0 ± 0%       ~ (p=1.000 n=10) ¹
    geomean                            ²               +0.00%                ²
    ¹ all samples are equal
    ² summaries must be >0 to compute geomean
```

### CODE REVIEW

**Comment by fyfyrchik** [comment]
CI benchmarks agree (about the same ~20% improvement):
```
BenchmarkLog10/Log10/uint256-36                           969393              1241 ns/op               0 B/op          0 allocs/op
```

From #136 
```
BenchmarkLog10/Log10/uint256-36                     	  705776	      1597 ns/op	       0 B/op	       0 allocs/op
```

File: uint256.go [hunk 1]
```diff
@@ -1358,33 +1355,23 @@ var (
...
+	bitlen := z.BitLen()
+	if bitlen == 0 {
+		return 0
 	}
-	// Non-uint64 space
-	if z.Cmp(&pows[res-20]) < 0 {
-		return uint(res - 1)
+
+	t := (bitlen + 1) * 1233 >> 12
+	if t < 20 && bitlen <= 64 && z[0] < pows64[t] || 20 <= t && z.Lt(&pows[t-20]) {
```
Comment (holiman) [inline uint256.go:1373]:
```suggestion
	if bitlen <= 64 && z[0] < pows64[t] || t >= 20 && z.Lt(&pows[t-20]) {
```

Removing the `t < 20` shaves off even more
```
name                   old time/op    new time/op    delta
Log10/Log10/uint256-8    2.41µs ±42%    1.90µs ±48%  -21.06%  (p=0.045 n=10+10)
```

Comment (fyfyrchik) [reply uint256.go]: Fixed.

File: uint256.go [hunk 2]
```diff
@@ -1358,33 +1355,23 @@ var (
...
+	//
+	// The idea is that log10(z) = log2(z) / log2(10)
+	// log2(z) trivially is z.Bitlen()
+	// 1/log2(10) is a constant ~ 1233 / 4096. The approximation is correct up to 5 digit after
+	// the decimal point and it seems no further refinement is needed.
+	// Our tests check all boundary cases anyway.
+
+	bitlen := z.BitLen()
+	if bitlen == 0 {
+		return 0
```
[the remark this answers is not in the document]
Comment (fyfyrchik) [reply uint256.go:1369]: Or we may check `t != 0` case in the main `if`, what do you think?

File: uint256.go (hunk 1 above)
Comment (fyfyrchik) [reply uint256.go]: I missed you added your commit, picked up your improvements too.

File: uint256.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (holiman) [reply uint256.go]: > The function comment says we return 0 for 0

Doesn't it? I thought it did

[the remark this answers is not in the document]
Comment (holiman) [reply uint256.go]: > We don't really need this special case, since the code below it works equally well for the zero-case.

That statement was wrong. I don't know, might be better to put it back the way it was. That big clause is complicated enough as it is :)

[the remark this answers is not in the document]
Comment (fyfyrchik) [reply uint256.go:1369]: Ok, fixed.

**Review by holiman — APPROVED** [review]
Thanks for this!
